src/replication.c

```diff
@@ -1112,15 +1113,65 @@ void restartAOFAfterSYNC() {
     }
 }
 
+long long getMonotonicTimeMs() {
+    struct timespec res;
+    clock_gettime(CLOCK_MONOTONIC, &res);
+    return (res.tv_sec * 1000000000LL + res.tv_nsec) / 1000000LL;
+}
+
+#define SLAVE_SYNC_BUDGET_UPDATE_INTERVAL_MS 10
+#define SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS 50
+void updateSlaveSyncBudget(ssize_t nread) {
+    if (server.slave_sync_speed_bps <= 0) return;
+    server.slave_sync_budget -= nread * 8000LL;
+
+    long long now_ms = getMonotonicTimeMs();
+
+    long long delta_time_ms = now_ms - server.slave_sync_budget_time_ms;
+    if (delta_time_ms > SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS) {
+        delta_time_ms = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS;
+    }
+    long long max_budget = SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS * server.slave_sync_speed_bps;
+    server.slave_sync_budget_time_ms = now_ms;
+    server.slave_sync_budget += delta_time_ms * server.slave_sync_speed_bps;
+    if (server.slave_sync_budget > max_budget) {
+        server.slave_sync_budget = max_budget;
+    }
+}
+
+
+int slaveSyncBudgetHandler(aeEventLoop *el, long long id, void *clientData) {
```

Comment:
Making a note of this since I was initially confused: So if I understand this correctly,

This limits the amount of data that will be `read()` from the tcp socket as described, and the read will acknowledge that data .

TCP connections have a limit on the amount of unacknowledged data there can be at any point in time (e.g. windows/limits on read/write memory) - once that fills up, then rate limiting the calls to read on the replica will effectively limit how much the redis master can send over the network to this replica, which should avoid performance degradation of other redis clients for the redis master

This PR seems like it should be effective in solving that issue